Python: Compute parquet stats

maxdebayser opened the pull request:
@Fokko 

This commit partly addresses issue https://github.com/apache/iceberg/issues/7256. Unfortunately the pyarrow library is not as flexible as we would like. When passing write_statistics=True to `pyarrow.parquet.write_table` the statistics are written out for each row group in the file, instead of computed globally.

In the issue a "metadata_collector" was mentioned which I assume is the parameter of the `pyarrow.parquet.write_metadata` function. The `pyarrow.parquet.write_table` function has no such parameter.

The function in this PR intentionally works at the level of individual parquet files instead of the dataset to support scenarios such as writing from Ray where each file of the dataset is written by a different task.

Fokko commented:
It looks like the metadata is also available on the plain writer, where we don't have to write a dataset:

```python
➜  Desktop python3                       
Python 3.11.3 (main, Apr  7 2023, 20:13:31) [Clang 14.0.0 (clang-1400.0.29.202)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import pyarrow as pa
>>> table = pa.table({'n_legs': [2, 2, 4, 4, 5, 100],
...                   'animal': ["Flamingo", "Parrot", "Dog", "Horse", "Brittle stars", "Centipede"]})
>>> import pyarrow.parquet as pq
>>> 
>>> writer = pq.ParquetWriter('/tmp/vo.parquet', table.schema)
>>> writer.write_table(table)
>>> writer.close()
>>> writer.writer.metadata.row_group(0)
<pyarrow._parquet.RowGroupMetaData object at 0x1375b9080>
  num_columns: 2
  num_rows: 6
  total_byte_size: 256
>>> writer.writer.metadata.row_group(0).to_dict()
{
	'num_columns': 2,
	'num_rows': 6,
	'total_byte_size': 256,
	'columns': [{
		'file_offset': 119,
		'file_path': '',
		'physical_type': 'INT64',
		'num_values': 6,
		'path_in_schema': 'n_legs',
		'is_stats_set': True,
		'statistics': {
			'has_min_max': True,
			'min': 2,
			'max': 100,
			'null_count': 0,
			'distinct_count': 0,
			'num_values': 6,
			'physical_type': 'INT64'
		},
		'compression': 'SNAPPY',
		'encodings': ('RLE_DICTIONARY', 'PLAIN', 'RLE'),
		'has_dictionary_page': True,
		'dictionary_page_offset': 4,
		'data_page_offset': 46,
		'total_compressed_size': 115,
		'total_uncompressed_size': 117
	}, {
		'file_offset': 359,
		'file_path': '',
		'physical_type': 'BYTE_ARRAY',
		'num_values': 6,
		'path_in_schema': 'animal',
		'is_stats_set': True,
		'statistics': {
			'has_min_max': True,
			'min': 'Brittle stars',
			'max': 'Parrot',
			'null_count': 0,
			'distinct_count': 0,
			'num_values': 6,
			'physical_type': 'BYTE_ARRAY'
		},
		'compression': 'SNAPPY',
		'encodings': ('RLE_DICTIONARY', 'PLAIN', 'RLE'),
		'has_dictionary_page': True,
		'dictionary_page_offset': 215,
		'data_page_offset': 302,
		'total_compressed_size': 144,
		'total_uncompressed_size': 139
	}]
}
```

maxdebayser commented:
@Fokko, I understand your concern, I think it's because we have different use cases in mind.

If I understand correctly you want to write a pyarrow.Table to a partitioned dataset with write_dataset. Therefore computing min/max on the whole Table is not what you need because you actually need the min/max for the columns of the individual files. (Just pointing out that with the metadata collector you get the stats for the row chunks, so you'll still have to compute the stats for the file from those).

I'm coming from a different use case. I would like to write from Ray using something like https://docs.ray.io/en/latest/data/api/doc/ray.data.Dataset.write_parquet.html#ray.data.Dataset.write_parquet . In this case there is no global pyarrow.Table that represent the dataset, Pyarrow tables are the blocks of the dataset that each individual ray task sees, for example in `map_batches`. In this scenario the pyarrow.write_dataset cannot be used because the full dataset is not entirely loaded into the memory of any compute node. In this scenario the GIL is also not a big concern because ray uses multiple worker processes.

I think we have to see if there is a way to have a single API for both use cases or if we'll need to have different API calls for both. In the second case it would be better to share part of the implementation to ensure that the behavior is consistent, but it could perhaps lead to bad performance.

Regarding the efficiency, the pyarrow.compute.min function is implemented in C++, so I think the performance is probably not a huge concern here. But I can try to compare both approaches with a large enough data set to measure it.

maxdebayser commented:
> It looks like the metadata is also available on the plain writer, where we don't have to write a dataset:
> 
> ```python
> ➜  Desktop python3                       
> Python 3.11.3 (main, Apr  7 2023, 20:13:31) [Clang 14.0.0 (clang-1400.0.29.202)] on darwin
> Type "help", "copyright", "credits" or "license" for more information.
> >>> import pyarrow as pa
> >>> table = pa.table({'n_legs': [2, 2, 4, 4, 5, 100],
> ...                   'animal': ["Flamingo", "Parrot", "Dog", "Horse", "Brittle stars", "Centipede"]})
> >>> import pyarrow.parquet as pq
> >>> 
> >>> writer = pq.ParquetWriter('/tmp/vo.parquet', table.schema)
> >>> writer.write_table(table)
> >>> writer.close()
> >>> writer.writer.metadata.row_group(0)
> <pyarrow._parquet.RowGroupMetaData object at 0x1375b9080>
>   num_columns: 2
>   num_rows: 6
>   total_byte_size: 256
> >>> writer.writer.metadata.row_group(0).to_dict()
> {
> 	'num_columns': 2,
> 	'num_rows': 6,
> 	'total_byte_size': 256,
> 	'columns': [{
> 		'file_offset': 119,
> 		'file_path': '',
> 		'physical_type': 'INT64',
> 		'num_values': 6,
> 		'path_in_schema': 'n_legs',
> 		'is_stats_set': True,
> 		'statistics': {
> 			'has_min_max': True,
> 			'min': 2,
> 			'max': 100,
> 			'null_count': 0,
> 			'distinct_count': 0,
> 			'num_values': 6,
> 			'physical_type': 'INT64'
> 		},
> 		'compression': 'SNAPPY',
> 		'encodings': ('RLE_DICTIONARY', 'PLAIN', 'RLE'),
> 		'has_dictionary_page': True,
> 		'dictionary_page_offset': 4,
> 		'data_page_offset': 46,
> 		'total_compressed_size': 115,
> 		'total_uncompressed_size': 117
> 	}, {
> 		'file_offset': 359,
> 		'file_path': '',
> 		'physical_type': 'BYTE_ARRAY',
> 		'num_values': 6,
> 		'path_in_schema': 'animal',
> 		'is_stats_set': True,
> 		'statistics': {
> 			'has_min_max': True,
> 			'min': 'Brittle stars',
> 			'max': 'Parrot',
> 			'null_count': 0,
> 			'distinct_count': 0,
> 			'num_values': 6,
> 			'physical_type': 'BYTE_ARRAY'
> 		},
> 		'compression': 'SNAPPY',
> 		'encodings': ('RLE_DICTIONARY', 'PLAIN', 'RLE'),
> 		'has_dictionary_page': True,
> 		'dictionary_page_offset': 215,
> 		'data_page_offset': 302,
> 		'total_compressed_size': 144,
> 		'total_uncompressed_size': 139
> 	}]
> }
> ```

Thanks for pointing that out, I completely missed the fact that you can get this data from the ParquetWriter after caling `close()`. That changes everything.

maxdebayser commented:
@Fokko , I've rewritten the implementation to use the pyarrow metadata_collector.  I've added a test cases to make sure that it's compatible with `write_dataset` as well as `ParquetWriter`. I've also also added a function to map the path of columns in the parquet metadata to Iceberg field IDs.

Fokko commented:
@maxdebayser can you rebase against lastest master?

maxdebayser commented:
@Fokko done!

Fokko commented on python/pyiceberg/utils/file_stats.py:
```diff
@@ -0,0 +1,333 @@
+
+
+def bool_to_avro(value: bool) -> bytes:
+    return struct.pack("?", value)
```
Can we initialize the structs just once? Similar to the Avro reading:

https://github.com/apache/iceberg/blob/e389e4d139624a49729379acd330dd9c96187b04/python/pyiceberg/avro/__init__.py#L19-L20

maxdebayser replied:
Sure

Fokko commented on python/pyiceberg/utils/file_stats.py:
```diff
@@ -0,0 +1,333 @@
+
+
+class StatsAggregator:
+    def __init__(self, type_string: str):
```
Why are we using a `type_string` here? The PyIceberg `PrimitiveType` seems to do the trick for me. We have methods for converting a PyArrow type to an IcebergType.

Nit:
```suggestion
    def __init__(self, type_string: str) -> None:
```

Fokko commented on python/pyiceberg/utils/file_stats.py:
```diff
@@ -0,0 +1,333 @@
+
+
+def fill_parquet_file_metadata(
+    df: DataFile, metadata: pq.FileMetaData, col_path_2_iceberg_id: Dict[str, int], file_size: int
```
Should we move this one to `pyarrow.py`? This uses PyArrow classes that might not be installed.

maxdebayser replied:
Thanks for pointing that out. I've moved the code over.

Fokko commented on python/tests/utils/test_file_stats.py:
```diff
@@ -0,0 +1,361 @@
+
+def construct_test_table() -> pa.Buffer:
+    schema = pa.schema(
+        [pa.field("strings", pa.string()), pa.field("floats", pa.float64()), pa.field("list", pa.list_(pa.int64()))]
```
Can we also add a map here?

maxdebayser replied:
Sure

Fokko commented on python/pyiceberg/utils/file_stats.py:
```diff
@@ -0,0 +1,333 @@
+    Args:
+        df (DataFile): A DataFile object representing the Parquet file for which metadata is to be filled.
+        metadata (pyarrow.parquet.FileMetaData): A pyarrow metadata object.
+        col_path_2_iceberg_id: A mapping of column paths as in the `path_in_schema` attribute of the colum
```
A suggestion. Since we have the Iceberg write schema, we could also easily construct a `List[int, int]` that will tell the mapping of position to field-id. We have the `PreOrderSchemaVisitor` where we traverse the schema in order to construct this list. I don't like the `key_value` specific to PyArrow, and the list will be faster. WDYT?

maxdebayser replied:
Yes, this is why I've added the `parquet_schema_to_ids` function that uses the  SchemaVisitor. If you prefer we can pass the schema to the `fill_parquet_file_metadata` and compute the mapping internally instead of making the caller of the function do it. The reason I decomposed it was to make testing easier.

The `parquet_schema_to_ids` uses the `path_in_schema` column metadata attribute instead of the position because I'm not 100% sure if I can assume anything about the position, whereas using the full column path should be safe.

Fokko replied:
Since Arrow is column-oriented, I'm sure that it will follow the order of the write schema:

![image](https://github.com/apache/iceberg/assets/1134248/5dd5016e-7761-4cdc-a169-d3c4744c581b)

The reason I try to avoid using too many internal details from PyArrow is that we support PyArrow `[9.0.0, 12.0.1]` currently. There is no guarantee that all these internals stay the same, therefore I think we should do as much as possible within our own control (also regarding the internal Parquet types, or how the column names are structured).

We already have the write schema, so we can easily filter out the primitive types:

```python
class PyArrowStatisticsCollector(PreOrderSchemaVisitor[List[StatisticsCollector]]):
    _field_id = 0
    _schema: Schema
    _properties: Properties

    def __init__(self, schema: Schema, properties: Dict[str, str]):
        self._schema = schema
        self._properties = properties

    def schema(self, schema: Schema, struct_result: Callable[[], List[StatisticsCollector]]) -> List[StatisticsCollector]:
        return struct_result()

    def struct(self, struct: StructType, field_results: List[Callable[[], List[StatisticsCollector]]]) -> List[StatisticsCollector]:
        return list(chain(*[result() for result in field_results]))

    def field(self, field: NestedField, field_result: Callable[[], List[StatisticsCollector]]) -> List[StatisticsCollector]:
        self._field_id = field.field_id
        result = field_result()
        return result

    def list(self, list_type: ListType, element_result: Callable[[], List[StatisticsCollector]]) -> List[StatisticsCollector]:
        self._field_id = list_type.element_id
        return element_result()

    def map(self, map_type: MapType, key_result: Callable[[], List[StatisticsCollector]], value_result: Callable[[], List[StatisticsCollector]]) -> List[StatisticsCollector]:
        self._field_id = map_type.key_id
        k = key_result()
        self._field_id = map_type.value_id
        v = value_result()
        return k + v

    def primitive(self, primitive: PrimitiveType) -> List[StatisticsCollector]:
        return [StatisticsCollector(
            field_id=self._field_id,
            iceberg_type=primitive,
            mode=MetricsMode.TRUNC
            # schema
            # self._properties.get(f"write.metadata.metrics.column.{schema.find_column_name(self._field_id)}")
        )]
```

This way we get a nice list of columns that we need to collect statistics for. We have:

```python
@dataclass(frozen=True)
class StatisticsCollector:
    field_id: int
    iceberg_type: PrimitiveType
    mode: MetricsMode
```

Where we can use the `field_id` to properly populate the maps, and the `iceberg_type` to feed into `to_bytes` to do the conversion (so we don't have to have yet another one for PyArrow).

I did a quick test, and it seems to work:
```python
def test_complex_schema(table_schema_nested: Schema):
    tbl = pa.Table.from_pydict({
        "foo": ["a", "b"],
        "bar": [1, 2],
        "baz": [False, True],
        "qux": [["a", "b"], ["c", "d"]],
        "quux": [[("a", (("aa", 1), ("ab", 2)))], [("b", (("ba", 3), ("bb", 4)))]],
        "location": [[(52.377956, 4.897070), (4.897070, -122.431297)],
                     [(43.618881, -116.215019), (41.881832, -87.623177)]],
        "person": [("Fokko", 33), ("Max", 42)]  # Possible data quality issue
    },
        schema=schema_to_pyarrow(table_schema_nested)
    )
    stats_columns = pre_order_visit(table_schema_nested, PyArrowStatisticsCollector(table_schema_nested, {}))

    visited_paths = []

    def file_visitor(written_file: Any) -> None:
        visited_paths.append(written_file)

    with TemporaryDirectory() as tmpdir:
        pq.write_to_dataset(tbl, tmpdir, file_visitor=file_visitor)

    assert visited_paths[0].metadata.num_columns == len(stats_columns)
```

Fokko commented on python/pyiceberg/utils/file_stats.py:
```diff
@@ -0,0 +1,333 @@
+                        col_aggs[col_id].add_max(statistics.max)
+
+                except pyarrow.lib.ArrowNotImplementedError:
+                    pass
```
Should we log a warning here?

Fokko commented on python/pyiceberg/utils/file_stats.py:
```diff
@@ -0,0 +1,333 @@
+                        if col_id not in col_aggs:
+                            col_aggs[col_id] = StatsAggregator(statistics.physical_type)
+
+                        col_aggs[col_id].add_min(statistics.min)
```
Do we need the intermediate `col_aggs`? I would prefer to directly write it into `{lower,upper}_bounds`.

Fokko replied:
Fair point, thanks for the context

Fokko commented on python/pyiceberg/utils/file_stats.py:
```diff
@@ -0,0 +1,333 @@
+            raise AssertionError(f"Unknown physical type {type_string}")
+
+    def add_min(self, val: bytes) -> None:
+        if not self.current_min:
```
```suggestion
        if self.current_min is not None:
```

Fokko commented on python/pyiceberg/utils/file_stats.py:
```diff
@@ -0,0 +1,333 @@
+            self.current_min = val
+
+    def add_max(self, val: bytes) -> None:
+        if not self.current_max:
```
```suggestion
        if self.current_max is not None:
```

Fokko commented on python/pyiceberg/utils/file_stats.py:
```diff
@@ -0,0 +1,333 @@
+    def add_min(self, val: bytes) -> None:
+        if not self.current_min:
+            self.current_min = val
+        elif val < self.current_min:
```
Any reason to not use Python's build in `min` function? That one might be passed down to C.

Fokko reviewed (commented):
I left some comments to generalize it, so we can also re-use it outside of PyArrow.

maxdebayser commented on python/pyiceberg/utils/file_stats.py:
```diff
@@ -0,0 +1,333 @@
+    StructType,
+)
+
+BOUND_TRUNCATED_LENGHT = 16
```
Thanks for pointing me in the right direction here, I've added a feature to configure this via table properties as documented here: https://iceberg.apache.org/docs/latest/configuration/

Fokko replied:
This is awesome, thank you

Fokko commented on python/pyiceberg/io/pyarrow.py:
```diff
@@ -1013,3 +1027,359 @@ def map_key_partner(self, partner_map: Optional[pa.Array]) -> Optional[pa.Array]
+#
+
+
+def bool_to_avro(value: bool) -> bytes:
```
When we use the Iceberg types, then we can re-use `to_bytes`:

https://github.com/apache/iceberg/blob/62cb7b5c50451085d6e59f0594292bf883af4220/python/pyiceberg/conversions.py#L154-L166

Fokko commented on python/tests/io/test_pyarrow.py:
```diff
@@ -1330,3 +1341,485 @@ def test_pyarrow_wrap_fsspec(example_task: FileScanTask, table_schema_simple: Sc
+        visited_paths.append(written_file)
+
+    with TemporaryDirectory() as tmpdir:
+        pq.write_to_dataset(table, tmpdir, partition_cols=["even"], file_visitor=file_visitor)
```
Let's leave out partitioning for now. We first have to investigate how to handle applying the partition transforms.

Fokko commented on python/pyiceberg/io/pyarrow.py:
```diff
@@ -1013,3 +1027,359 @@ def map_key_partner(self, partner_map: Optional[pa.Array]) -> Optional[pa.Array]
+
+
+BOUND_TRUNCATED_LENGHT = 16
+TRUNCATION_EXPR = r"^truncate\((\d+)\)$"
```
Oh, that's a great catch, missed that one. Very sad :'(

maxdebayser commented on python/pyiceberg/io/pyarrow.py:
```diff
@@ -1013,3 +1027,359 @@ def map_key_partner(self, partner_map: Optional[pa.Array]) -> Optional[pa.Array]
+class MetricsMode(Enum):
+    NONE = 0
+    COUNTS = 1
+    TRUNC = 2
```
I think an Enum is a more pythonic way to define a set of constants.

maxdebayser replied:
Actually now I'm seeing how it would fit into the StatisticsCollector visitor. It makes sense, let me implement your suggestion.

Fokko commented:
For some reason my big comment was collapsed:

Since Arrow is column-oriented, I'm sure that it will follow the order of the write schema:

![image](https://github.com/apache/iceberg/assets/1134248/5dd5016e-7761-4cdc-a169-d3c4744c581b)

The reason I try to avoid using too many internal details from PyArrow is that we support PyArrow `[9.0.0, 12.0.1]` currently. There is no guarantee that all these internals stay the same, therefore I think we should do as much as possible within our own control (also regarding the internal Parquet types, or how the column names are structured).

We already have the write schema, so we can easily filter out the primitive types:

```python
class PyArrowStatisticsCollector(PreOrderSchemaVisitor[List[StatisticsCollector]]):
    _field_id = 0
    _schema: Schema
    _properties: Properties

    def __init__(self, schema: Schema, properties: Dict[str, str]):
        self._schema = schema
        self._properties = properties

    def schema(self, schema: Schema, struct_result: Callable[[], List[StatisticsCollector]]) -> List[StatisticsCollector]:
        return struct_result()

    def struct(self, struct: StructType, field_results: List[Callable[[], List[StatisticsCollector]]]) -> List[StatisticsCollector]:
        return list(chain(*[result() for result in field_results]))

    def field(self, field: NestedField, field_result: Callable[[], List[StatisticsCollector]]) -> List[StatisticsCollector]:
        self._field_id = field.field_id
        result = field_result()
        return result

    def list(self, list_type: ListType, element_result: Callable[[], List[StatisticsCollector]]) -> List[StatisticsCollector]:
        self._field_id = list_type.element_id
        return element_result()

    def map(self, map_type: MapType, key_result: Callable[[], List[StatisticsCollector]], value_result: Callable[[], List[StatisticsCollector]]) -> List[StatisticsCollector]:
        self._field_id = map_type.key_id
        k = key_result()
        self._field_id = map_type.value_id
        v = value_result()
        return k + v

    def primitive(self, primitive: PrimitiveType) -> List[StatisticsCollector]:
        return [StatisticsCollector(
            field_id=self._field_id,
            iceberg_type=primitive,
            mode=MetricsMode.TRUNC
            # schema
            # self._properties.get(f"write.metadata.metrics.column.{schema.find_column_name(self._field_id)}")
        )]
```

This way we get a nice list of columns that we need to collect statistics for. We have:

```python
@dataclass(frozen=True)
class StatisticsCollector:
    field_id: int
    iceberg_type: PrimitiveType
    mode: MetricsMode
```

Where we can use the `field_id` to properly populate the maps, and the `iceberg_type` to feed into `to_bytes` to do the conversion (so we don't have to have yet another one for PyArrow).

I did a quick test, and it seems to work:
```python
def test_complex_schema(table_schema_nested: Schema):
    tbl = pa.Table.from_pydict({
        "foo": ["a", "b"],
        "bar": [1, 2],
        "baz": [False, True],
        "qux": [["a", "b"], ["c", "d"]],
        "quux": [[("a", (("aa", 1), ("ab", 2)))], [("b", (("ba", 3), ("bb", 4)))]],
        "location": [[(52.377956, 4.897070), (4.897070, -122.431297)],
                     [(43.618881, -116.215019), (41.881832, -87.623177)]],
        "person": [("Fokko", 33), ("Max", 42)]  # Possible data quality issue
    },
        schema=schema_to_pyarrow(table_schema_nested)
    )
    stats_columns = pre_order_visit(table_schema_nested, PyArrowStatisticsCollector(table_schema_nested, {}))

    visited_paths = []

    def file_visitor(written_file: Any) -> None:
        visited_paths.append(written_file)

    with TemporaryDirectory() as tmpdir:
        pq.write_to_dataset(tbl, tmpdir, file_visitor=file_visitor)

    assert visited_paths[0].metadata.num_columns == len(stats_columns)
```

maxdebayser commented on python/pyiceberg/io/pyarrow.py:
```diff
@@ -1013,3 +1027,359 @@ def map_key_partner(self, partner_map: Optional[pa.Array]) -> Optional[pa.Array]
+            trunc_lengths = {n: t for n in col_names}
+
+        for col_name in col_names:
+            col_mode = table_metadata.properties.get(f"write.metadata.metrics.column.{col_name}")
```
I couldn't find a lot of good examples of the Java API, so I'm assuming that the `Schema.find_column_name()` returns a fully qualified column name that is consistent with the Java API.

Fokko commented on python/pyiceberg/io/pyarrow.py:
```diff
@@ -1013,3 +1025,271 @@ def map_key_partner(self, partner_map: Optional[pa.Array]) -> Optional[pa.Array]
+
+            value_counts[col_id] = value_counts.get(col_id, 0) + column.num_values
+
+            if column.is_stats_set:
```
When this isn't set, we probably want to emit a `logger.warn("PyArrow statistics missing when writing file")`

rdblue commented on python/pyiceberg/io/pyarrow.py:
```diff
@@ -1013,3 +1025,271 @@ def map_key_partner(self, partner_map: Optional[pa.Array]) -> Optional[pa.Array]
+        return self.serialize(self.current_min)[: self.trunc_length]
+
+    def get_max(self) -> bytes:
+        return self.serialize(self.current_max)[: self.trunc_length]
```
Truncation is more complicated because you can corrupt the max by truncating. We'll need to implement the equivalent of `UnicodeUtil` for max.

maxdebayser replied:
Thanks for pointing that out. I've just read the UnicodeUtil code and it seems to me that the truncation length there is interpreted as length in characters and not in bytes. Is that understanding correct?

And should primitive types be truncated at all? I don't think it makes a lot of sense.

Fokko replied:
Strings in Python are already utf8, so we could truncate them before converting them to bytes. I don't think we need to truncate int/float/double/etc.

rdblue commented on python/pyiceberg/io/pyarrow.py:
```diff
@@ -1013,3 +1025,271 @@ def map_key_partner(self, partner_map: Optional[pa.Array]) -> Optional[pa.Array]
+
+
+DEFAULT_METRICS_MODE_KEY = "write.metadata.metrics.default"
+COLUMN_METRICS_MODE_KEY = "write.metadata.metrics.column"
```
This is a prefix, right?

maxdebayser replied:
Yes, I've changed the name to reflect that.

Fokko commented on python/pyiceberg/io/pyarrow.py:
```diff
@@ -1013,3 +1025,271 @@ def map_key_partner(self, partner_map: Optional[pa.Array]) -> Optional[pa.Array]
+        else:
+            split_offsets.append(data_offset)
+
+        for c in range(parquet_metadata.num_columns):
+            col_id = stats_columns[c].field_id
+
+            column = row_group.column(c)
+
+            column_sizes[col_id] = column_sizes.get(col_id, 0) + column.total_compressed_size
+
+            metrics_mode = stats_columns[c].mode
+
+            if metrics_mode == MetricsMode(MetricModeTypes.NONE):
+                continue
+
+            value_counts[col_id] = value_counts.get(col_id, 0) + column.num_values
+
+            if column.is_stats_set:
+                try:
+                    statistics = column.statistics
+
+                    null_value_counts[col_id] = null_value_counts.get(col_id, 0) + statistics.null_count
+
+                    if metrics_mode == MetricsMode(MetricModeTypes.COUNTS):
+                        continue
+
+                    if col_id not in col_aggs:
+                        col_aggs[col_id] = StatsAggregator(statistics.physical_type, metrics_mode.length)
+
+                    col_aggs[col_id].add_min(statistics.min)
+                    col_aggs[col_id].add_max(statistics.max)
```
What do you think of the following? This way we have fewer lookups. Also, I think it is best to stick with the Iceberg naming (field-id over column-id).
```suggestion
        for pos, stats_col in enumerate(stats_columns):
            field_id = stats_col.field_id

            column = row_group.column(pos)

            column_sizes[field_id] = column_sizes.get(field_id, 0) + column.total_compressed_size

            if stats_col.mode == MetricsMode(MetricModeTypes.NONE):
                continue

            value_counts[field_id] = value_counts.get(field_id, 0) + column.num_values

            if column.is_stats_set:
                try:
                    statistics = column.statistics

                    null_value_counts[field_id] = null_value_counts.get(field_id, 0) + statistics.null_count

                    if stats_col.mode == MetricsMode(MetricModeTypes.COUNTS):
                        continue

                    if field_id not in col_aggs:
                        col_aggs[field_id] = StatsAggregator(statistics.physical_type, stats_col.mode.length)

                    col_aggs[field_id].add_min(statistics.min)
                    col_aggs[field_id].add_max(statistics.max)
```

Fokko commented on python/pyiceberg/io/pyarrow.py:
```diff
@@ -1013,3 +1025,271 @@ def map_key_partner(self, partner_map: Optional[pa.Array]) -> Optional[pa.Array]
         return partner_map.items if isinstance(partner_map, pa.MapArray) else None
+
+
+class StatsAggregator:
+    def __init__(self, type_string: str, trunc_length: Optional[int] = None) -> None:
+        self.current_min: Any = None
+        self.current_max: Any = None
+        self.trunc_length = trunc_length
+        self.primitive_type: Optional[PrimitiveType] = None
+
+        if type_string == "BOOLEAN":
+            self.primitive_type = BooleanType()
+        elif type_string == "INT32":
+            self.primitive_type = IntegerType()
+        elif type_string == "INT64":
+            self.primitive_type = LongType()
+        elif type_string == "INT96":
+            raise NotImplementedError("Statistics not implemented for INT96 physical type")
+        elif type_string == "FLOAT":
+            self.primitive_type = FloatType()
+        elif type_string == "DOUBLE":
+            self.primitive_type = DoubleType()
+        elif type_string == "BYTE_ARRAY":
+            self.primitive_type = BinaryType()
+        elif type_string == "FIXED_LEN_BYTE_ARRAY":
+            self.primitive_type = BinaryType()
+        else:
+            raise AssertionError(f"Unknown physical type {type_string}")
```
```suggestion
class StatsAggregator:
    current_min: Any
    current_max: Any
    trunc_length: Optional[int]
    primitive_type: PrimitiveType

    def __init__(self, iceberg_type: PrimitiveType, trunc_length: Optional[int] = None) -> None:
        self.current_min: Any = None
        self.current_max: Any = None
        self.trunc_length = trunc_length
        self.primitive_type: PrimitiveType = iceberg_type
```
Now we can also remove the `assert`'s to appease mypy.

maxdebayser replied:
I agree that the code can be simplified like that. I just want to run a thought by you: since we have this redundancy in type information should we perhaps take advantage of this to check that the types match? In this way we would be able to detect bugs that result in the data written to parquet being of a different type than what is in the schema.

Fokko reviewed (commented):
Getting close, a few more suggestions

Fokko commented on python/tests/io/test_pyarrow.py:
```diff
@@ -1330,3 +1342,411 @@ def test_pyarrow_wrap_fsspec(example_task: FileScanTask, table_schema_simple: Sc
+    assert datafile.split_offsets is not None
+    assert len(datafile.split_offsets) == 1
+    assert datafile.split_offsets[0] == 4
+
```
Thanks for all the tests, this is great @maxdebayser!

Could you add these two for my sanity:

```python
def test_write_and_read_stats_schema(table_schema_nested: Schema):
    tbl = pa.Table.from_pydict({
        "foo": ["a", "b"],
        "bar": [1, 2],
        "baz": [False, True],
        "qux": [["a", "b"], ["c", "d"]],
        "quux": [[("a", (("aa", 1), ("ab", 2)))], [("b", (("ba", 3), ("bb", 4)))]],
        "location": [[(52.377956, 4.897070), (4.897070, -122.431297)],
                     [(43.618881, -116.215019), (41.881832, -87.623177)]],
        "person": [("Fokko", 33), ("Max", 42)]  # Possible data quality issue
    },
        schema=schema_to_pyarrow(table_schema_nested)
    )
    stats_columns = pre_order_visit(table_schema_nested, PyArrowStatisticsCollector(table_schema_nested, {}))

    visited_paths = []

    def file_visitor(written_file: Any) -> None:
        visited_paths.append(written_file)

    with tempfile.TemporaryDirectory() as tmpdir:
        pq.write_to_dataset(tbl, tmpdir, file_visitor=file_visitor)

    assert visited_paths[0].metadata.num_columns == len(stats_columns)

def test_stats_types(table_schema_nested: Schema):
    stats_columns = pre_order_visit(table_schema_nested, PyArrowStatisticsCollector(table_schema_nested, {}))

    # the field-ids should be sorted
    assert all(stats_columns[i].field_id <= stats_columns[i + 1].field_id for i in range(len(stats_columns)-1))
    assert [col.iceberg_type for col in stats_columns] == [
        StringType(),
        IntegerType(),
        BooleanType(),
        StringType(),
        StringType(),
        StringType(),
        IntegerType(),
        FloatType(),
        FloatType(),
        StringType(),
        IntegerType(),
    ]

```

Fokko reviewed (commented):
I forgot to publish the comments...

maxdebayser commented on python/pyiceberg/io/pyarrow.py:
```diff
@@ -1025,3 +1040,291 @@ def map_key_partner(self, partner_map: Optional[pa.Array]) -> Optional[pa.Array]
+        self.primitive_type = iceberg_type
+
+    def serialize(self, value: Any) -> bytes:
+        if type(value) == date:
```
Since this commit already has test cases that depend on this conversion, can we address the refactoring in that issue?

Fokko replied:
Yes, that was my intention. Let's try to get this in ASAP

maxdebayser replied:
What I meant was can we merge it like this and then refactor in the other issue?

Fokko replied:
Yes, let's refactor in a separate PR.

Fokko replied:
The PR is in :)

Fokko commented on python/pyiceberg/io/pyarrow.py:
```diff
@@ -1025,3 +1040,337 @@ def map_key_partner(self, partner_map: Optional[pa.Array]) -> Optional[pa.Array]
+        return to_bytes(self.primitive_type, value)
+
+    def add_min(self, val: Any) -> None:
+        if self.current_min is None:
+            self.current_min = val
+        else:
+            self.current_min = min(val, self.current_min)
```
```suggestion
        self.current_min = val if self.current_min is None else min(val, self.current_min)
```

Fokko commented on python/pyiceberg/io/pyarrow.py:
```diff
@@ -1025,3 +1040,337 @@ def map_key_partner(self, partner_map: Optional[pa.Array]) -> Optional[pa.Array]
+            self.current_min = min(val, self.current_min)
+
+    def add_max(self, val: Any) -> None:
+        if self.current_max is None:
+            self.current_max = val
+        else:
+            self.current_max = max(self.current_max, val)
```
```suggestion
        self.current_max = val if self.current_max is None else max(val, self.current_max)
```

Fokko commented on python/pyiceberg/io/pyarrow.py:
```diff
@@ -1025,3 +1040,337 @@ def map_key_partner(self, partner_map: Optional[pa.Array]) -> Optional[pa.Array]
+            self.current_max = max(self.current_max, val)
+
+    def get_min(self) -> bytes:
+        if self.primitive_type == StringType():
+            if type(self.current_min) != str:
+                raise ValueError("Expected the current_min to be a string")
+            return self.serialize(self.current_min[: self.trunc_length])
+        else:
+            return self.serialize(self.current_min)[: self.trunc_length]
```
```suggestion
        return self.serialize(
            self.current_min
            if self.trunc_length is None
            else TruncateTransform(width=self.trunc_length).transform(self.primitive_type)(self.current_min)
        )
```

Fokko commented on python/pyiceberg/io/pyarrow.py:
```diff
@@ -1025,3 +1040,337 @@ def map_key_partner(self, partner_map: Optional[pa.Array]) -> Optional[pa.Array]
+            return self.serialize(self.current_max)[: self.trunc_length]
+
+
+DEFAULT_TRUNCATION_LENGHT = 16
```
```suggestion
DEFAULT_TRUNCATION_LENGTH = 16
```

Fokko commented on python/pyiceberg/io/pyarrow.py:
```diff
@@ -1025,3 +1040,337 @@ def map_key_partner(self, partner_map: Optional[pa.Array]) -> Optional[pa.Array]
+        if column_name is None:
+            raise ValueError(f"Column for field {self._field_id} not found")
+
+        metrics_mode = MetricsMode(MetricModeTypes.TRUNCATE, DEFAULT_TRUNCATION_LENGHT)
```
```suggestion
        metrics_mode = MetricsMode(MetricModeTypes.TRUNCATE, DEFAULT_TRUNCATION_LENGTH)
```

Fokko reviewed (commented):
Last few comments from my end, @rdblue can you do a final pass to make sure we haven't overlooked anything?

rdblue commented on python/pyiceberg/utils/file_stats.py:
```diff
@@ -0,0 +1,164 @@
+            from file_object using pyarrow.parquet.read_table.
+    """
+    
+    parquet_file = pq.ParquetFile(file_object)
```
I completely agree here. We do not want to collect stats by reading individual data files. We should be able to get them when writing the file.

maxdebayser replied:
Yes, collecting the statistics while writing has been implemented.

rdblue commented on python/pyiceberg/io/pyarrow.py:
```diff
@@ -1025,3 +1039,335 @@ def map_key_partner(self, partner_map: Optional[pa.Array]) -> Optional[pa.Array]
+
+    def serialize(self, value: Any) -> bytes:
+        if self.primitive_type == UUIDType():
+            value = uuid.UUID(bytes=value)
```
Is this needed because Arrow internally represents the UUID type as `fixed[16]`? If so, can't we just return the value directly since it's already bytes?

Fokko replied:
Yes, this will be done in https://github.com/apache/iceberg/pull/8267 including testing

Fokko replied:
@maxdebayser The PR is in, can you remove this check?

We can now replace serialize in the constructor with:
```python
from functools import partial
self.serialize: Callable[[Any], bytes] = partial(to_bytes, self.primitive_type)
```

rdblue commented on python/pyiceberg/io/pyarrow.py:
```diff
@@ -1025,3 +1039,335 @@ def map_key_partner(self, partner_map: Optional[pa.Array]) -> Optional[pa.Array]
+    def add_max(self, val: Any) -> None:
+        self.current_max = val if self.current_max is None else max(val, self.current_max)
+
+    def get_min(self) -> bytes:
```
Style: unnecessary `get` in a method name.

rdblue commented on python/pyiceberg/io/pyarrow.py:
```diff
@@ -1025,3 +1039,335 @@ def map_key_partner(self, partner_map: Optional[pa.Array]) -> Optional[pa.Array]
+
+        return to_bytes(self.primitive_type, value)
+
+    def add_min(self, val: Any) -> None:
```
I'm not sure `add` makes sense here. What about something like `update_min`?

Also, in cases where this is used to accumulate a min and max from a list of values, should there be an `update` method that delegates to both `update_min` and `update_max`?

rdblue commented on python/pyiceberg/io/pyarrow.py:
```diff
@@ -1025,3 +1039,335 @@ def map_key_partner(self, partner_map: Optional[pa.Array]) -> Optional[pa.Array]
+        return self.serialize(
+            self.current_min
+            if self.trunc_length is None
+            else TruncateTransform(width=self.trunc_length).transform(self.primitive_type)(self.current_min)
```
It seems odd to create a transform here instead of in the constructor, but I guess if this is only ever called once?

maxdebayser replied:
Yes, it's called only once.

rdblue commented on python/pyiceberg/io/pyarrow.py:
```diff
@@ -1025,3 +1039,335 @@ def map_key_partner(self, partner_map: Optional[pa.Array]) -> Optional[pa.Array]
+            if type(self.current_max) != str:
+                raise ValueError("Expected the current_max to be a string")
+
+            s_result = self.current_max[: self.trunc_length]
```
I don't think this is a good place to put this logic. I would prefer adding it in a separate PR with dedicated tests for the util method that produces a truncated upper bound.

rdblue commented on python/pyiceberg/io/pyarrow.py:
```diff
@@ -1025,3 +1039,335 @@ def map_key_partner(self, partner_map: Optional[pa.Array]) -> Optional[pa.Array]
+
+            return self.serialize(b_result)
+        else:
+            return self.serialize(self.current_max)[: self.trunc_length]
```
How could this logic be correct? It seems suspicious to me. I think that it's likely that the idea here was that this will never be called, but instead this should throw an exception or return None to signal the value cannot be truncated.

rdblue commented on python/pyiceberg/io/pyarrow.py:
```diff
@@ -1025,3 +1039,335 @@ def map_key_partner(self, partner_map: Optional[pa.Array]) -> Optional[pa.Array]
+
+
+DEFAULT_TRUNCATION_LENGTH = 16
+TRUNCATION_EXPR = r"^truncate\((\d+)\)$"
```
Since this is strict with string start and end, we should make sure that `trim` is called to remove accidental whitespace.

maxdebayser replied:
I'm trying to follow the behavior of the Java codebase, and I couldn't find a trim being used anywhere here  https://github.com/apache/iceberg/blob/6b1c9f0a398235b56977e817eb5d3d9457c160b5/core/src/main/java/org/apache/iceberg/MetricsModes.java#L39C21-L39C21 and here https://github.com/apache/iceberg/blob/6b1c9f0a398235b56977e817eb5d3d9457c160b5/core/src/main/java/org/apache/iceberg/MetricsConfig.java#L194

But we can add it if you prefer.

rdblue commented on python/pyiceberg/io/pyarrow.py:
```diff
@@ -1025,3 +1039,335 @@ def map_key_partner(self, partner_map: Optional[pa.Array]) -> Optional[pa.Array]
+
+
+def match_metrics_mode(mode: str) -> MetricsMode:
+    sanitized_mode = mode.lower()
```
And trim?

rdblue commented on python/pyiceberg/io/pyarrow.py:
```diff
@@ -1025,3 +1039,335 @@ def map_key_partner(self, partner_map: Optional[pa.Array]) -> Optional[pa.Array]
+def match_metrics_mode(mode: str) -> MetricsMode:
+    sanitized_mode = mode.lower()
+    if sanitized_mode.startswith("truncate"):
+        m = re.match(TRUNCATION_EXPR, mode, re.IGNORECASE)
```
Why `re.IGNORECASE` when using `sanitized_mode` that has been converted to lower?

rdblue commented on python/pyiceberg/io/pyarrow.py:
```diff
@@ -1025,3 +1039,335 @@ def map_key_partner(self, partner_map: Optional[pa.Array]) -> Optional[pa.Array]
+            return MetricsMode(MetricModeTypes.TRUNCATE, int(m[1]))
+        else:
+            raise ValueError(f"Malformed truncate: {mode}")
+    elif sanitized_mode.startswith("none"):
```
Why does this permit additional characters when truncate is strict?

maxdebayser replied:
It didn't, it's an artifact of the review process. The first version of the code matched beginning and end of the string:
```
def match_metrics_mode(mode: str) -> MetricsMode:
    m = re.match(TRUNCATION_EXPR, mode, re.IGNORECASE)
    if m:
        length = int(m[1])
        if length < 1:
            raise ValueError("Truncation length must be larger than 0")
        return MetricsMode(MetricModeTypes.TRUNCATE, int(m[1]))
    elif re.match("^none$", mode, re.IGNORECASE):
        return MetricsMode(MetricModeTypes.NONE)
    elif re.match("^counts$", mode, re.IGNORECASE):
        return MetricsMode(MetricModeTypes.COUNTS)
    elif re.match("^full$", mode, re.IGNORECASE):
        return MetricsMode(MetricModeTypes.FULL)
    else:
        raise ValueError(f"Unsupported metrics mode {mode}")
```

rdblue commented on python/pyiceberg/io/pyarrow.py:
```diff
@@ -1025,3 +1039,335 @@ def map_key_partner(self, partner_map: Optional[pa.Array]) -> Optional[pa.Array]
+        return list(chain(*[result() for result in field_results]))
+
+    def field(self, field: NestedField, field_result: Callable[[], List[StatisticsCollector]]) -> List[StatisticsCollector]:
+        self._field_id = field.field_id
```
I don't like relying on pre-order traversal for keeping the field ID. This is exactly what the `before_field` and `after_field` callbacks are intended for in the normal `SchemaVisitor`, and using that doesn't require separate logic for map key/value and list element fields.

The main issue is that the pre-order visitor doesn't actually perform a pre-order visit. It creates `Callable` results that the visitor can use to perform a pre-order visit. I think that makes this more complex, harder to use, and more likely to break.

rdblue commented on python/pyiceberg/io/pyarrow.py:
```diff
@@ -1025,3 +1039,335 @@ def map_key_partner(self, partner_map: Optional[pa.Array]) -> Optional[pa.Array]
+    def primitive(self, primitive: PrimitiveType) -> List[StatisticsCollector]:
+        column_name = self._schema.find_column_name(self._field_id)
+        if column_name is None:
+            raise ValueError(f"Column for field {self._field_id} not found")
```
Why not just skip creating a stats collector for this field?

rdblue commented on python/pyiceberg/io/pyarrow.py:
```diff
@@ -1025,3 +1039,335 @@ def map_key_partner(self, partner_map: Optional[pa.Array]) -> Optional[pa.Array]
+        if column_name is None:
+            raise ValueError(f"Column for field {self._field_id} not found")
+
+        metrics_mode = MetricsMode(MetricModeTypes.TRUNCATE, DEFAULT_TRUNCATION_LENGTH)
```
Should the lhs be a constant?

rdblue commented on python/pyiceberg/io/pyarrow.py:
```diff
@@ -1025,3 +1039,335 @@ def map_key_partner(self, partner_map: Optional[pa.Array]) -> Optional[pa.Array]
+
+        default_mode = self._properties.get(DEFAULT_METRICS_MODE_KEY)
+        if default_mode:
+            metrics_mode = match_metrics_mode(default_mode)
```
The default mode construction can be done in the constructor rather than in every leaf node, right?

rdblue commented on python/pyiceberg/io/pyarrow.py:
```diff
@@ -1025,3 +1039,335 @@ def map_key_partner(self, partner_map: Optional[pa.Array]) -> Optional[pa.Array]
+            compute the mapping if column position to iceberg schema type id. It's also used to set the mode
+            for column metrics collection
+    """
+    schema = next(filter(lambda s: s.schema_id == table_metadata.current_schema_id, table_metadata.schemas))
```
This work should not be done by the caller. The `TableMetadata` should either provide a map by `schema_id` or should have a method that returns a `Schema` by ID.

maxdebayser replied:
I agree, but I don't think it's in the scope of this PR.

rdblue commented on python/pyiceberg/io/pyarrow.py:
```diff
@@ -1025,3 +1039,335 @@ def map_key_partner(self, partner_map: Optional[pa.Array]) -> Optional[pa.Array]
+
+    stats_columns = pre_order_visit(schema, PyArrowStatisticsCollector(schema, table_metadata.properties))
+
+    if parquet_metadata.num_columns != len(stats_columns):
```
I think it should be fine to omit some metadata if it can't be calculated. I'd rather have metadata for some columns than only for all columns.

Fokko replied:
I think we should leave this one in, for now, I think they are always the same, but when this is not the case, then we should be notified

rdblue commented on python/pyiceberg/io/pyarrow.py:
```diff
@@ -1025,3 +1039,335 @@ def map_key_partner(self, partner_map: Optional[pa.Array]) -> Optional[pa.Array]
+        for pos, stats_col in enumerate(stats_columns):
+            field_id = stats_col.field_id
+
+            column = row_group.column(pos)
```
I don't think it is a good idea to match by position. There is no guarantee that the current schema of the table still matches the write schema used for this data file. Even making assumptions about how this method will be called, this is dangerous.

Instead, I think this should index the columns by `path_in_schema`:

```python
columns_by_path = { row_group.column(pos).path_in_schema: row_group.column(pos) for pos in range(row_group.num_columns) }
```

Then this code can get the metrics for each column using `StatisticsCollector.column_name` to find the stats in the map.

Fokko replied:
This would need some translation because PyArrow uses different naming:

![image](https://github.com/apache/iceberg/assets/1134248/f2d194fc-4ba5-4c0d-82bd-bee87da1737d)

I haven't found any way to get PyArrow out of order.

rdblue commented on python/pyiceberg/io/pyarrow.py:
```diff
@@ -1025,3 +1039,335 @@ def map_key_partner(self, partner_map: Optional[pa.Array]) -> Optional[pa.Array]
+    df: DataFile,
+    parquet_metadata: pq.FileMetaData,
+    file_size: int,
+    table_metadata: TableMetadata,
```
I don't think that this should be passed in. All that's needed is the Parquet write schema and a map from field ID to metrics mode. There are good reasons not to require that bundled into `TableMetadata`:
1. It is safe and possible to write new data files with any older schema, so the current table schema may not be the write schema (this would currently break in that case)
2. Metrics modes can be pre-calculated to produce a map from field ID to mode. It makes more sense to do this in a common util method than to mix the code into the already complex `StatisticsCollector` visitor
3. Avoiding passing `TableMetadata` keeps this logic cleaner and forces some separation of concerns

maxdebayser replied:
Yes, the first iteration had this separation of concerns: https://github.com/apache/iceberg/blob/ce8a5df5d07b51dbee03c7d8d88dcff99ffbde02/python/pyiceberg/utils/file_stats.py#L121 . We can split it up again.

Fokko replied:
I agree that we shouldn't pass in the `TableMetadata` here. Since we know the write schema, we can just pass that one in.

> Metrics modes can be pre-calculated to produce a map from field ID to mode. It makes more sense to do this in a common util method than to mix the code into the already complex StatisticsCollector visitor

I'm not sure about this one, we know the order of the columns, so we can just rely on that one. We can also have a lookup, but I'm not sure that it makes it much simpler than keeping the mode next to the id and type.

Fokko replied:
We could also just pass in the `stats_columns`:
```suggestion
    stats_columns: List[StatisticsCollector]
```

rdblue commented on python/pyiceberg/io/pyarrow.py:
```diff
@@ -1025,3 +1039,335 @@ def map_key_partner(self, partner_map: Optional[pa.Array]) -> Optional[pa.Array]
+
+            column = row_group.column(pos)
+
+            column_sizes[field_id] = column_sizes.get(field_id, 0) + column.total_compressed_size
```
Minor: seems cleaner to me to use `setdefault`:

```python
column_sizes.setdefault(field_id, 0)
column_sizes[field_id] += column.total_compressed_size
```

rdblue commented on python/pyiceberg/io/pyarrow.py:
```diff
@@ -1025,3 +1039,335 @@ def map_key_partner(self, partner_map: Optional[pa.Array]) -> Optional[pa.Array]
+                except pyarrow.lib.ArrowNotImplementedError as e:
+                    logger.warning(e)
+            else:
+                logger.warning("PyArrow statistics missing for column %d when writing file", pos)
```
If this (or an exception above) happens, then this needs to ensure that the lower and upper bounds are discarded if they are present for other row groups. In Java, this is done by [tracking field IDs](https://github.com/apache/iceberg/blob/master/parquet/src/main/java/org/apache/iceberg/parquet/ParquetUtil.java#L137) for which at least one row group didn't have stats, then suppressing min/max later on.

Otherwise, you could be missing stats for a row group and not know what the actual min/max is, but still produce lower and upper bounds based on the data that was actually there.

maxdebayser replied:
Thanks. It's always better to protect ourselves against bugs in other libraries.

rdblue commented on python/pyiceberg/io/pyarrow.py:
```diff
@@ -1025,3 +1039,335 @@ def map_key_partner(self, partner_map: Optional[pa.Array]) -> Optional[pa.Array]
+    upper_bounds = {}
+
+    for k, agg in col_aggs.items():
+        lower_bounds[k] = agg.get_min()
```
Instead of `get_min`, it would be better to signal that this is serializing the value. What about `min_as_bytes`?

rdblue replied:
Min could also be None, in which case it should not be written.

rdblue commented on python/pyiceberg/io/pyarrow.py:
```diff
@@ -1025,3 +1039,335 @@ def map_key_partner(self, partner_map: Optional[pa.Array]) -> Optional[pa.Array]
+        if _max is not None:
+            upper_bounds[k] = _max
+
+    df.file_format = FileFormat.PARQUET
```
@Fokko, are we okay with `DataFile` instances being mutable?

Fokko replied:
No, I prefer to keep them immutable, but I'm okay with it for now

rdblue commented on python/tests/io/test_pyarrow.py:
```diff
@@ -16,16 +16,39 @@
 import tempfile
-from typing import Any, List, Optional
+import uuid
+from datetime import (
```
@Fokko, why are there `datetime` classes here?

I think that tests like these should completely avoid Python datetime logic and instead should directly use the internal representations.

Fokko replied:
We use this for PyArrow which has the datetime classes in their public API.

rdblue commented on python/tests/io/test_pyarrow.py:
```diff
@@ -1345,3 +1374,655 @@ def test_pyarrow_wrap_fsspec(example_task: FileScanTask, table_schema_simple: Sc
     )
+
+
+def construct_test_table() -> Tuple[Any, Any, Union[TableMetadataV1, TableMetadataV2]]:
```
I think the tests for Parquet stats should go into a separate test file. There's too much in `test_pyarrow` as it is and this is a reasonably self-contained set of tests.

Fokko replied:
`test_pyarrow_stats.py`?

rdblue commented on python/tests/io/test_pyarrow.py:
```diff
@@ -1345,3 +1374,655 @@ def test_pyarrow_wrap_fsspec(example_task: FileScanTask, table_schema_simple: Sc
+    assert datafile.value_counts[1] == 4
+    assert datafile.value_counts[2] == 4
+    assert datafile.value_counts[5] == 10  # 3 lists with 3 items and a None value
+    assert datafile.value_counts[6] == 5
```
It looks like both maps and lists count empty/null parent structures. I would expect this to be the number of leaf values.

Can you confirm that this is also the behavior of the Java Parquet library? If not, we should just drop map and list stats because they aren't the same.

Fokko replied:
With Spark:

```sql
CREATE TABLE nyc.test_map_maps2 AS 
SELECT map_from_arrays(array(1.0, 3.0), array('2', '4')) as map, array('a', 'b', 'c') as arr
```

Schema:
```json
{
	"type": "struct",
	"schema-id": 0,
	"fields": [{
		"id": 1,
		"name": "map",
		"required": false,
		"type": {
			"type": "map",
			"key-id": 3,
			"key": "decimal(2, 1)",
			"value-id": 4,
			"value": "string",
			"value-required": false
		}
	}, {
		"id": 2,
		"name": "arr",
		"required": false,
		"type": {
			"type": "list",
			"element-id": 5,
			"element": "string",
			"element-required": false
		}
	}]
}
```

We don't get any stats in Spark:
```json
{
	"status": 1,
	"snapshot_id": {
		"long": 4895801649705337905
	},
	"data_file": {
		"file_path": "s3://warehouse/nyc/test_map_maps/data/00000-95-750d8f3e-8d49-44ec-b37e-9e101e003a5d-00001.parquet",
		"file_format": "PARQUET",
		"partition": {},
		"record_count": 1,
		"file_size_in_bytes": 1438,
		"block_size_in_bytes": 67108864,
		"column_sizes": {
			"array": [{
				"key": 3,
				"value": 57
			}, {
				"key": 4,
				"value": 58
			}, {
				"key": 5,
				"value": 61
			}]
		},
		"value_counts": {
			"array": [{
				"key": 3,
				"value": 2
			}, {
				"key": 4,
				"value": 2
			}, {
				"key": 5,
				"value": 3
			}]
		},
		"null_value_counts": {
			"array": [{
				"key": 3,
				"value": 0
			}, {
				"key": 4,
				"value": 0
			}, {
				"key": 5,
				"value": 0
			}]
		},
		"nan_value_counts": {
			"array": []
		},
		"lower_bounds": {
			"array": []
		},
		"upper_bounds": {
			"array": []
		},
		"key_metadata": null,
		"split_offsets": {
			"array": [4]
		},
		"sort_order_id": {
			"int": 0
		}
	}
}
```

The stats are only computed for the primitive types (see `PyArrowStatisticsCollector`).

Fokko replied:
```sql
CREATE TABLE default.arrays
SELECT
    array(1, 2, 3, null) as with_a_null,
    array(1, 2, cast('NaN' as double), null) as with_null_and_nan,
    array(1, 2, 3, cast('NaN' as double)) as with_a_nan
```

Schema:
```json
{
	"type": "struct",
	"schema-id": 0,
	"fields": [{
		"id": 1,
		"name": "with_a_null",
		"required": false,
		"type": {
			"type": "list",
			"element-id": 4,
			"element": "int",
			"element-required": false
		}
	}, {
		"id": 2,
		"name": "with_null_and_nan",
		"required": false,
		"type": {
			"type": "list",
			"element-id": 5,
			"element": "double",
			"element-required": false
		}
	}, {
		"id": 3,
		"name": "with_a_nan",
		"required": false,
		"type": {
			"type": "list",
			"element-id": 6,
			"element": "double",
			"element-required": false
		}
	}]
}
```

The missing upper- and lower bound seems to be an issue with Spark. It does generates the doubles:

```json
{
	"status": 1,
	"snapshot_id": {
		"long": 3911973389432551915
	},
	"data_file": {
		"file_path": "s3://warehouse/default/arrays/data/00000-0-1e0d4a20-4473-450e-9b1e-9196b5edac66-00001.parquet",
		"file_format": "PARQUET",
		"partition": {},
		"record_count": 1,
		"file_size_in_bytes": 1310,
		"block_size_in_bytes": 67108864,
		"column_sizes": {
			"array": [{
				"key": 4,
				"value": 60
			}, {
				"key": 5,
				"value": 63
			}, {
				"key": 6,
				"value": 66
			}]
		},
		"value_counts": {
			"array": [{
				"key": 4,
				"value": 4
			}, {
				"key": 5,
				"value": 4
			}, {
				"key": 6,
				"value": 4
			}]
		},
		"null_value_counts": {
			"array": [{
				"key": 4,
				"value": 1
			}, {
				"key": 5,
				"value": 1
			}, {
				"key": 6,
				"value": 0
			}]
		},
		"nan_value_counts": {
			"array": [{
				"key": 5,
				"value": 1
			}, {
				"key": 6,
				"value": 1
			}]
		},
		"lower_bounds": {
			"array": [{
				"key": 5,
				"value": "\u0000\u0000\u0000\u0000\u0000\u0000ð?"
			}, {
				"key": 6,
				"value": "\u0000\u0000\u0000\u0000\u0000\u0000ð?"
			}]
		},
		"upper_bounds": {
			"array": [{
				"key": 5,
				"value": "\u0000\u0000\u0000\u0000\u0000\u0000\u0000@"
			}, {
				"key": 6,
				"value": "\u0000\u0000\u0000\u0000\u0000\u0000\b@"
			}]
		},
		"key_metadata": null,
		"split_offsets": {
			"array": [4]
		},
		"sort_order_id": {
			"int": 0
		}
	}
}
```

Fokko replied:
Spark has the same number of values:
![image](https://github.com/apache/iceberg/assets/1134248/ad3a7f71-2e7d-46af-8309-c2720b97ff75)

maxdebayser replied:
Seems to be consistent with what we're seeing in python

rdblue commented on python/tests/io/test_pyarrow.py:
```diff
@@ -1345,3 +1374,655 @@ def test_pyarrow_wrap_fsspec(example_task: FileScanTask, table_schema_simple: Sc
+        {},
+    ]
+
+    table = pa.Table.from_pydict(
```
I think that this should have a nested struct without a list or map as well.

rdblue commented on python/tests/io/test_pyarrow.py:
```diff
@@ -1345,3 +1374,655 @@ def test_pyarrow_wrap_fsspec(example_task: FileScanTask, table_schema_simple: Sc
+    fill_parquet_file_metadata(datafile, metadata, len(file_bytes), table_metadata)
+
+    assert len(datafile.column_sizes) == 5
+    # these values are an artifact of how the write_table encodes the columns
```
Do we want to test these directly or just assert that the value is > 0?

rdblue commented on python/tests/io/test_pyarrow.py:
```diff
@@ -1345,3 +1374,655 @@ def test_pyarrow_wrap_fsspec(example_task: FileScanTask, table_schema_simple: Sc
+
+    # #arrow does not include this in the statistics
+    # assert len(datafile.nan_value_counts)  == 3
+    # assert datafile.nan_value_counts[1]    == 0
```
These should be `None` for any column that is not a float or double type.

Fokko replied:
I was just wondering. The evaluator is not taking those into consideration for any not `float` or `double` type, any reason to still add those?

maxdebayser replied:
You're talking about the nan_value_counts, right? Pyarrow==12.0.1 doesn't compute this statistic unfortunately.

Fokko replied:
Opened an issue a while ago: https://github.com/apache/arrow/issues/36068

rdblue commented on python/tests/io/test_pyarrow.py:
```diff
@@ -1345,3 +1374,655 @@ def test_pyarrow_wrap_fsspec(example_task: FileScanTask, table_schema_simple: Sc
+    assert len(datafile.null_value_counts) == 5
+    assert datafile.null_value_counts[1] == 1
+    assert datafile.null_value_counts[2] == 0
+    assert datafile.null_value_counts[5] == 1
```
I also think these values are incorrect. The `null` value is not in a leaf, so it should not be reported for id=5.

Fokko replied:
Type:
```json
{"type": "list", "element-id": 5, "element": "long", "element-required": False}
```
And the data:
```python
_list = [[1, 2, 3], [4, 5, 6], None, [7, 8, 9]]
```
I count a single `null`

maxdebayser replied:
This is how pyarrow is returning these values. Unless we look at the actual table it seems to me that the only options are not returning this statistic or returning it as is.
```
<pyarrow._parquet.ColumnChunkMetaData object at 0x7fbd4a8a15e0>
  file_offset: 543
  file_path: 
  physical_type: INT64
  num_values: 10
  path_in_schema: list.list.element
  is_stats_set: True
  statistics:
    <pyarrow._parquet.Statistics object at 0x7fbd4a8a1130>
      has_min_max: True
      min: 1
      max: 9
      null_count: 1
      distinct_count: 0
      num_values: 9
      physical_type: INT64
      logical_type: None
      converted_type (legacy): NONE
  compression: SNAPPY
  encodings: ('RLE_DICTIONARY', 'PLAIN', 'RLE')
  has_dictionary_page: True
  dictionary_page_offset: 392
  data_page_offset: 455
  total_compressed_size: 151
  total_uncompressed_size: 174
```

rdblue commented on python/tests/io/test_pyarrow.py:
```diff
@@ -1345,3 +1374,655 @@ def test_pyarrow_wrap_fsspec(example_task: FileScanTask, table_schema_simple: Sc
+
+    assert len(datafile.lower_bounds) == 3
+    assert datafile.lower_bounds[2] == STRUCT_FLOAT.pack(1.69)
+    assert 5 not in datafile.lower_bounds
```
This doesn't look right. Integers and longs are not affected by truncate. For non-string or binary types, the mode should be equivalent to `full`.

rdblue reviewed (changes requested):
Looks like a good design overall, but I pointed out a few areas where I think stats are wrong (probably from PyArrow) and I think that collection should be a bit more resilient.

Fokko commented on python/pyiceberg/io/pyarrow.py:
```diff
@@ -1025,3 +1039,344 @@ def map_key_partner(self, partner_map: Optional[pa.Array]) -> Optional[pa.Array]
+        return [StatisticsCollector(field_id=self._field_id, iceberg_type=primitive, mode=metrics_mode, column_name=column_name)]
+
+
+def fill_parquet_file_metadata(
```
Let's make sure that this isn't being used outside of the module.
```suggestion
def _fill_parquet_file_metadata(
```

Fokko commented:
@maxdebayser Thanks for working on this, and for your patience because it took quite a while, and we also went a bit back and forth. Let's merge this in. As a next step, we can revisit https://github.com/apache/iceberg/pull/8012 and add a lot of integration tests to make sure that we don't have any correctness issues. 

I think there are still some open-ends on the statistics, but they are not limited to Python. Please refer to https://github.com/apache/iceberg/issues/8598

Fokko commented:
Thanks @rdblue for the review, that was very helpful.

maxdebayser commented:
Thanks, @Fokko and @rdblue
